tensorizer/serialization.py

```diff
@@ -0,0 +1,1071 @@
+# try to import UNIX only dependencies
+try:
+    import fcntl
+    import resource
+except ImportError:
+    fcntl = None
+    resource = None
+
+from enum import Enum
+import collections.abc
+import io
+import os
+import tensorizer.stream_io as stream_io
+import tensorizer.utils as utils
+import mmap
+import numpy
+import struct
+import time
+import torch
+import typing
+import logging
+import tempfile
+import hashlib
+import zlib
+import ctypes
+
+if torch.cuda.is_available():
+    cudart = torch.cuda.cudart()
+else:
+    cudart = None
+
+from collections import OrderedDict
+from typing import Optional, Tuple, Union, List, Iterator, Dict, Callable, Any
+
+lz4 = None
+
+__all__ = ["TensorSerializer", "TensorDeserializer"]
+
+# Setup logger
+logger = logging.getLogger(__name__)
+
+
+# Whether the tensor is a parameter or a buffer on the model.
+class TensorType(Enum):
+    PARAM = 0
+    BUFFER = 1
+    STATE_DICT = 2
+
+
+TENSORIZER_VERSION = 1
+TENSORIZER_MAGIC = b"|TZR|"
+
+
+class TensorEntry(typing.TypedDict):
+    name: str
+    type: TensorType
+    offset: int
+    data_offset: int
+    data_length: int
+    dtype: str
+    shape: List[int]
+
+
+class HashType(Enum):
+    CRC32 = 0
+    SHA256 = 1
+
+
+class TensorHash(typing.TypedDict):
+    type: HashType
+    hash: bytes
+
+
+class TensorDeserializer(collections.abc.Mapping):
+    """
+    Given a file-like object for read, deserialize tensors to a state_dict or
+    a torch.nn.Module.
+
+    Args:
+        file_obj: A file-like object to read from. It can also be a string
+            representing a path to a file or an HTTP/HTTPS/S3 URI.
+        device: The device to load the tensors to.
+        filter_func: A function (tensor_name: str) -> bool that returns True
+            if a tensor should be loaded, or False if it should be skipped.
+            If None, all tensors are loaded.
+        dtype: The dtype to load the tensors as. If None, the dtype will be
+            inferred from the file.
+        lazy_load: If True, tensors will be loaded and cached when keys are
+            accessed. If False, all tensors will be loaded into memory up
+            front.
+        plaid_mode: If True, tensors will be loaded extremely fast into the
+            target device. This is only supported on CUDA devices, and the
+            buffers are going to be inconsistent due to the extreme
+            naughtiness of reusing a backing buffer. This is only recommended
+            for use with inference, and not training.
+    """
+
+    def __init__(
+            self,
+            file_obj: Union[io.BufferedIOBase, io.RawIOBase,
+            typing.BinaryIO, str],
+            device: Union[torch.device, str, None] = None,
+            filter_func: Optional[Callable[[str], Union[bool, Any]]] = None,
+            dtype: Union[torch.dtype, str, None] = None,
+            *,
+            lazy_load: bool = False,
+            plaid_mode: bool = False):
+        if isinstance(file_obj, str):
```

Comment:
I would just use `Black` if we want to enforce formatting.  Repo could benefit by precommit rules and/or CI linting.

Reply:
We'd preferably go with Black's style. Right now we don't have all developers consistently using Black prior to commits, though, so it leads to slightly confusing diffs when one in every few pull requests changes a bunch of files for style, as it obscures the primary intent of the PR.

To fix that, we should add a pre-commit hook to format code with Black, so that all commits are already formatted coming in. However, since that doesn't affect the functionality of the package or break anything, I would say that we should probably work on that later, after the package releases. (We could also run Black at least once before the release).